synapse/handlers/federation_event.py

```diff
@@ -959,6 +973,27 @@ async def _get_state_after_missing_prev_event(
 
         return remote_state
 
+    async def _get_state_and_persist(
+        self, destination: str, room_id: str, event_id: str
+    ) -> None:
+        """Get the complete room state at a given event, and persist any new events
+        as outliers"""
+        room_version = await self._store.get_room_version(room_id)
+        auth_events, state_events = await self._federation_client.get_room_state(
+            destination, room_id, event_id=event_id, room_version=room_version
+        )
+        logger.info("/state returned %i events", len(auth_events) + len(state_events))
+
+        await self._auth_and_persist_outliers(
+            room_id, itertools.chain(auth_events, state_events)
+        )
+
+        # we also need the event itself.
+        if not await self._store.have_seen_events(room_id, event_id):
```

Comment:
I think you want

```suggestion
        if not await self._store.have_seen_events(room_id, (event_id,)):
```

Mypy can't spot this because `x: str` implies `x: Iterable[str]`. @squahtx noted this pain point a few weeks ago I think.

(Given that CI passed, are we missing some kind of test here?)

Reply:
Xref: https://github.com/python/typing/issues/256

Reply:
Good spot, thank you.

I've fixed this by implementing `have_seen_event` and using that instead, which seems to have a lower potential for foot-shooting.

I've also added a UT which catches this.

The failure mode here was relatively edge-casey (it requires us to *have* the event, but not its state), and benign (it resulted in a redundant call to `_matrix/federation/event`), which was why this wasn't being picked up in the tests before.